cin_validator/rules/cin2022_23/rule_4009.py

```diff
@@ -0,0 +1,268 @@
+    sample_cinplan["CINPlanEndDate"] = pd.to_datetime(
+        sample_cinplan["CINPlanEndDate"], format="%d/%m/%Y", errors="coerce"
+    )
+
+    # Run the rule function, passing in our sample data.
+    result = run_rule(
+        validate,
+        {
+            ChildIdentifiers: sample_ci,
+            CINplanDates: sample_cinplan,
+        },
+    )
+
+    # Use .type2_issues to check for the result of .push_type2_issues() which you used above.
+    issues_list = result.type2_issues
+    assert len(issues_list) == 2
+    # the function returns a list on NamedTuples where each NamedTuple contains (table, column_list, df_issues)
+    # pick any table and check it's values. the tuple in location 1 will contain the Reviews columns because that's the second thing pushed above.
+    issues = issues_list[1]
+
+    # get table name and check it. Replace Reviews with the name of your table.
+    issue_table = issues.table
+    assert issue_table == CINplanDates
+
+    # check that the right columns were returned. Replace CPPreviewDate  with a list of your columns.
+    issue_columns = issues.columns
+    assert issue_columns == [CINPlanEndDate]
+
+    # check that the location linking dataframe was formed properly.
+    issue_rows = issues.row_df
+    # replace 3 with the number of failing points you expect from the sample data.
+    assert len(issue_rows) == 3
+    # check that the failing locations are contained in a DataFrame having the appropriate columns. These lines do not change.
+    assert isinstance(issue_rows, pd.DataFrame)
+    assert issue_rows.columns.to_list() == ["ERROR_ID", "ROW_ID"]
+
+    # Create the dataframe which you expect, based on the fake data you created. It should have two columns.
+    # - The first column is ERROR_ID which contains the unique combination that identifies each error instance, which you decided on, in your zip, earlier.
+    # - The second column in ROW_ID which contains a list of index positions that belong to each error instance.
+
+    # The ROW ID values represent the index positions where you expect the sample data to fail the validation check.
+    expected_df = pd.DataFrame(
+        [
+            {
+                "ERROR_ID": (
+                    "child2",  # ChildID
+                    # CINPlanEndDate
+                    pd.to_datetime("29/06/2002", format="%d/%m/%Y", errors="coerce"),
+                ),
+                "ROW_ID": [1],
+            },
+            {
+                "ERROR_ID": (
+                    "child3",  # ChildID
+                    # CINPlanEndDate
+                    pd.to_datetime("07/02/2001", format="%d/%m/%Y", errors="coerce"),
+                ),
+                "ROW_ID": [2],
+            },
+            {
+                "ERROR_ID": (
+                    "child7",  # ChildID
+                    # CINPlanEndDate
+                    pd.to_datetime("14/03/2008", format="%d/%m/%Y", errors="coerce"),
+                ),
+                "ROW_ID": [6],
```

Comment:
Hi Tracy, you've used CINplanEndDates in this expected df ERROR_ID, but in your validate function, you've set the ERROR_ID to put out PersonDeathDate, so, you'll need to either update these to reflect that, or update the statement that makes the ERROR_ID in validate to reflect these.